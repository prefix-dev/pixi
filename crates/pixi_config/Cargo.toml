[package]
authors.workspace = true
description = "Contains the global configuration for use in a pixi project"
edition.workspace = true
homepage.workspace = true
license.workspace = true
name = "pixi_config"
readme.workspace = true
repository.workspace = true
version = "0.1.0"

[dependencies]
clap = { workspace = true, features = ["std", "derive", "env"] }
console = { workspace = true }
dirs = { workspace = true }
itertools = { workspace = true }
miette = { workspace = true }
pixi_consts = { workspace = true }
rattler = { workspace = true }
rattler_conda_types = { workspace = true }
rattler_repodata_gateway = { workspace = true, optional = true, features = [
  "gateway",
] }
reqwest-middleware = { workspace = true }
serde = { workspace = true }
serde_ignored = { workspace = true }
serde_json = { workspace = true }
toml_edit = { workspace = true, features = ["serde"] }
tracing = { workspace = true }
url = { workspace = true }

[dev-dependencies]
insta = { workspace = true, features = ["yaml"] }
rstest = { workspace = true }
