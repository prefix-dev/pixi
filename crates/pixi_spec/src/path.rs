use std::path::Path;

use rattler_conda_types::{package::ArchiveIdentifier, NamelessMatchSpec};
use typed_path::{Utf8NativePathBuf, Utf8TypedPathBuf};

use crate::SpecConversionError;

/// A specification of a package from a git repository.
#[derive(Debug, Clone, Hash, Eq, PartialEq)]
pub struct PathSpec {
    /// The path to the package
    pub path: Utf8TypedPathBuf,
}

impl PathSpec {
    /// Converts this instance into a [`NamelessMatchSpec`] if the path points
    /// to binary archive.
    pub fn try_into_nameless_match_spec(
        self,
        root_dir: &Path,
    ) -> Result<Option<NamelessMatchSpec>, SpecConversionError> {
        if !self.is_binary() {
            // Not a binary package
            return Ok(None);
        }

        // Convert the path to an absolute path based on the root_dir
        let path = if self.path.is_absolute() {
            self.path
        } else if let Ok(user_path) = self.path.strip_prefix("~/") {
            let home_dir = dirs::home_dir()
                .ok_or_else(|| SpecConversionError::InvalidPath(self.path.to_string()))?;
            let Some(home_dir_str) = home_dir.to_str() else {
                return Err(SpecConversionError::NotUtf8RootDir(home_dir));
            };
            Utf8TypedPathBuf::from(home_dir_str)
                .join(user_path)
                .normalize()
        } else {
            let Some(root_dir_str) = root_dir.to_str() else {
                return Err(SpecConversionError::NotUtf8RootDir(root_dir.to_path_buf()));
            };
            let native_root_dir = Utf8NativePathBuf::from(root_dir_str);
            if !native_root_dir.is_absolute() {
                return Err(SpecConversionError::NonAbsoluteRootDir(
                    root_dir.to_path_buf(),
                ));
            }

            native_root_dir.to_typed_path().join(self.path).normalize()
        };

        // Convert the absolute url to a file:// url
        let local_file_url =
            file_url::file_path_to_url(path.to_path()).expect("failed to convert path to file url");

        Ok(Some(NamelessMatchSpec {
            url: Some(local_file_url),
            ..NamelessMatchSpec::default()
        }))
    }

    /// Converts this instance into a [`PathSourceSpec`] if the path points to a
    /// source package. Otherwise, returns this instance unmodified.
    #[allow(clippy::result_large_err)]
    pub fn try_into_source_path(self) -> Result<PathSourceSpec, Self> {
        if self.is_binary() {
            Err(self)
        } else {
            Ok(PathSourceSpec { path: self.path })
        }
    }

    /// Returns true if this path points to a binary archive.
    pub fn is_binary(&self) -> bool {
        self.path
            .file_name()
            .and_then(ArchiveIdentifier::try_from_path)
            .is_some()
    }
}

/// Path to a source package. Different from [`PathSpec`] in that this type only
/// refers to source packages.
#[derive(Debug, Clone, Hash, Eq, PartialEq)]
pub struct PathSourceSpec {
    /// The path to the package. Either a directory or an archive.
    pub path: Utf8TypedPathBuf,
}

impl From<PathSourceSpec> for PathSpec {
    fn from(value: PathSourceSpec) -> Self {
        Self { path: value.path }
    }
}
