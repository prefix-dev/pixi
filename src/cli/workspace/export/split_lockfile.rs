use std::path::PathBuf;

use crate::{
    WorkspaceLocator,
    cli::cli_config::{LockFileUpdateConfig, WorkspaceConfig},
    lock_file::UpdateLockFileOptions,
};
use clap::Parser;
use miette::IntoDiagnostic;
use pixi_config::ConfigCli;
use rattler_lock::{Channel, DEFAULT_ENVIRONMENT_NAME, LockFile};

#[derive(Debug, Parser)]
#[clap(arg_required_else_help = false)]
pub struct Args {
    #[clap(flatten)]
    pub workspace_config: WorkspaceConfig,

    /// Output directory for each split lock file.
    /// The paths of the output lock files are: {output_dir}/{platform}/{environment}.lock
    pub output_dir: PathBuf,

    /// Keep the original environment name in the output lock file instead of replacing by "default"
    #[arg(long, default_value = "false")]
    pub keep_env_name: bool,

    #[clap(flatten)]
    pub lock_file_update_config: LockFileUpdateConfig,

    #[clap(flatten)]
    config: ConfigCli,
}

pub async fn execute(args: Args) -> miette::Result<()> {
    let workspace = WorkspaceLocator::for_cli()
        .with_search_start(args.workspace_config.workspace_locator_start())
        .locate()?
        .with_cli_config(args.config.clone());

    let lockfile = workspace
        .update_lock_file(UpdateLockFileOptions {
            lock_file_usage: args.lock_file_update_config.lock_file_usage()?,
            no_install: args.lock_file_update_config.no_lockfile_update,
            max_concurrent_solves: workspace.config().max_concurrent_solves(),
        })
        .await?;

    split(lockfile.as_lock_file(), args)
}

fn split(lockfile: &LockFile, args: Args) -> miette::Result<()> {
    if lockfile.is_empty() {
        eprintln!("lockfile is empty.");
        return Ok(());
    }

    let mut split_lockfiles = Vec::new();

    for (env_name, env) in lockfile.environments() {
        let output_env_name = if args.keep_env_name {
            env_name
        } else {
            DEFAULT_ENVIRONMENT_NAME
        };
        for plat in env.platforms() {
            let mut builder = LockFile::builder()
                .with_channels(output_env_name, env.channels().iter().map(Channel::clone))
                .with_options(output_env_name, env.solve_options().clone());
            if let Some(pypi_indexes) = env.pypi_indexes() {
                builder.set_pypi_indexes(output_env_name, pypi_indexes.clone());
            }
            if let Some(packages) = env.packages(plat) {
                for p in packages {
                    builder.add_package(output_env_name, plat, p.into());
                }
            }
            let s = builder.finish();
            if s.is_empty() {
                tracing::warn!("Ignore empty environment {env_name} on platform {plat}");
            } else {
                split_lockfiles.push((env_name.to_string(), plat, s));
            }
        }
    }

    if split_lockfiles.is_empty() {
        eprintln!("No environments.");
    } else {
        fs_err::create_dir_all(&args.output_dir).into_diagnostic()?;
        for (env_name, plat, l) in split_lockfiles {
            let subdir = &args.output_dir.join(plat.to_string());
            fs_err::create_dir_all(subdir).into_diagnostic()?;
            l.to_path(subdir.join(format!("{}.lock", env_name)).as_path())
                .into_diagnostic()?;
        }
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::Path;
    use tempfile::tempdir;

    #[test]
    fn test_split_lockfile() {
        let lockfile = Path::new(env!("CARGO_MANIFEST_DIR"))
            .join("tests/data/mock-projects/test-project-export/pixi.lock");

        let lockfile = LockFile::from_path(&lockfile).unwrap();

        let output_dir = tempdir().unwrap();

        let args = Args {
            workspace_config: WorkspaceConfig::default(),
            output_dir: output_dir.path().to_path_buf(),
            keep_env_name: false,
            lock_file_update_config: LockFileUpdateConfig::default(),
            config: ConfigCli::default(),
        };

        split(&lockfile, args).unwrap();

        for (env_name, env) in lockfile.environments() {
            for plat in env.platforms() {
                let output = output_dir
                    .path()
                    .join(plat.to_string())
                    .join(format!("{}.lock", env_name));
                let output_lockfile = LockFile::from_path(output.as_path()).unwrap();
                let snapshot = format!("test_split_lockfile_{}_{}", env_name, plat);
                insta::assert_snapshot!(snapshot, output_lockfile.render_to_string().unwrap());
            }
        }
    }
}
