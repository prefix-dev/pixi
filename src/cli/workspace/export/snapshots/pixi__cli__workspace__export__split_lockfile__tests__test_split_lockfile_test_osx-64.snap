---
source: src/cli/workspace/export/split_lockfile.rs
expression: output_lockfile.render_to_string().unwrap()
---
version: 6
environments:
  default:
    channels:
    - url: https://conda.anaconda.org/conda-forge/
    indexes:
    - https://pypi.org/simple
    packages:
      osx-64:
      - conda: https://conda.anaconda.org/conda-forge/osx-64/bzip2-1.0.8-hfdf4475_7.conda
      - conda: https://conda.anaconda.org/conda-forge/osx-64/ca-certificates-2024.8.30-h8857fd0_0.conda
      - conda: https://conda.anaconda.org/conda-forge/noarch/colorama-0.4.6-pyhd8ed1ab_0.tar.bz2
      - conda: https://conda.anaconda.org/conda-forge/noarch/exceptiongroup-1.2.2-pyhd8ed1ab_0.conda
      - conda: https://conda.anaconda.org/conda-forge/noarch/iniconfig-2.0.0-pyhd8ed1ab_0.conda
      - conda: https://conda.anaconda.org/conda-forge/osx-64/libexpat-2.6.2-h73e2aa4_0.conda
      - conda: https://conda.anaconda.org/conda-forge/osx-64/libffi-3.4.2-h0d85af4_5.tar.bz2
      - conda: https://conda.anaconda.org/conda-forge/osx-64/libsqlite-3.46.1-h4b8f8c9_0.conda
      - conda: https://conda.anaconda.org/conda-forge/osx-64/libzlib-1.3.1-h87427d6_1.conda
      - conda: https://conda.anaconda.org/conda-forge/osx-64/ncurses-6.5-hf036a51_1.conda
      - conda: https://conda.anaconda.org/conda-forge/osx-64/openssl-3.3.2-hd23fc13_0.conda
      - conda: https://conda.anaconda.org/conda-forge/noarch/packaging-24.1-pyhd8ed1ab_0.conda
      - conda: https://conda.anaconda.org/conda-forge/noarch/pluggy-1.5.0-pyhd8ed1ab_0.conda
      - conda: https://conda.anaconda.org/conda-forge/noarch/pytest-8.3.2-pyhd8ed1ab_0.conda
      - conda: https://conda.anaconda.org/conda-forge/osx-64/python-3.12.5-h37a9e06_0_cpython.conda
      - conda: https://conda.anaconda.org/conda-forge/osx-64/python_abi-3.12-5_cp312.conda
      - conda: https://conda.anaconda.org/conda-forge/osx-64/pyyaml-6.0.2-py312hb553811_1.conda
      - conda: https://conda.anaconda.org/conda-forge/osx-64/readline-8.2-h9e318b2_1.conda
      - conda: https://conda.anaconda.org/conda-forge/osx-64/tk-8.6.13-h1abcd95_1.conda
      - conda: https://conda.anaconda.org/conda-forge/noarch/tomli-2.0.1-pyhd8ed1ab_0.tar.bz2
      - conda: https://conda.anaconda.org/conda-forge/noarch/tzdata-2024a-h8827d51_1.conda
      - conda: https://conda.anaconda.org/conda-forge/osx-64/xz-5.2.6-h775f41a_0.tar.bz2
      - conda: https://conda.anaconda.org/conda-forge/osx-64/yaml-0.2.5-h0d85af4_2.tar.bz2
      - pypi: https://files.pythonhosted.org/packages/42/d7/1ec15b46af6af88f19b8e5ffea08fa375d433c998b8a7639e76935c14f1f/markdown_it_py-3.0.0-py3-none-any.whl
      - pypi: https://files.pythonhosted.org/packages/b3/38/89ba8ad64ae25be8de66a6d463314cf1eb366222074cfda9ee839c56a4b4/mdurl-0.1.2-py3-none-any.whl
      - pypi: https://files.pythonhosted.org/packages/f7/3f/01c8b82017c199075f8f788d0d906b9ffbbc5a47dc9918a945e13d5a2bda/pygments-2.18.0-py3-none-any.whl
      - pypi: https://files.pythonhosted.org/packages/c7/d9/c2a126eeae791e90ea099d05cb0515feea3688474b978343f3cdcfe04523/rich-13.8.0-py3-none-any.whl
packages:
- conda: https://conda.anaconda.org/conda-forge/osx-64/bzip2-1.0.8-hfdf4475_7.conda
  sha256: cad153608b81fb24fc8c509357daa9ae4e49dfc535b2cb49b91e23dbd68fc3c5
  md5: 7ed4301d437b59045be7e051a0308211
  depends:
  - __osx >=10.13
  license: bzip2-1.0.6
  license_family: BSD
  purls: []
  size: 134188
  timestamp: 1720974491916
- conda: https://conda.anaconda.org/conda-forge/osx-64/ca-certificates-2024.8.30-h8857fd0_0.conda
  sha256: 593f302d0f44c2c771e1614ee6d56fffdc7d616e6f187669c8b0e34ffce3e1ae
  md5: b7e5424e7f06547a903d28e4651dbb21
  license: ISC
  purls: []
  size: 158665
  timestamp: 1725019059295
- conda: https://conda.anaconda.org/conda-forge/noarch/colorama-0.4.6-pyhd8ed1ab_0.tar.bz2
  sha256: 2c1b2e9755ce3102bca8d69e8f26e4f087ece73f50418186aee7c74bef8e1698
  md5: 3faab06a954c2a04039983f2c4a50d99
  depends:
  - python >=3.7
  license: BSD-3-Clause
  license_family: BSD
  purls:
  - pkg:pypi/colorama?source=hash-mapping
  size: 25170
  timestamp: 1666700778190
- conda: https://conda.anaconda.org/conda-forge/noarch/exceptiongroup-1.2.2-pyhd8ed1ab_0.conda
  sha256: e0edd30c4b7144406bb4da975e6bb97d6bc9c0e999aa4efe66ae108cada5d5b5
  md5: d02ae936e42063ca46af6cdad2dbd1e0
  depends:
  - python >=3.7
  license: MIT and PSF-2.0
  purls:
  - pkg:pypi/exceptiongroup?source=hash-mapping
  size: 20418
  timestamp: 1720869435725
- conda: https://conda.anaconda.org/conda-forge/noarch/iniconfig-2.0.0-pyhd8ed1ab_0.conda
  sha256: 38740c939b668b36a50ef455b077e8015b8c9cf89860d421b3fff86048f49666
  md5: f800d2da156d08e289b14e87e43c1ae5
  depends:
  - python >=3.7
  license: MIT
  license_family: MIT
  purls:
  - pkg:pypi/iniconfig?source=hash-mapping
  size: 11101
  timestamp: 1673103208955
- conda: https://conda.anaconda.org/conda-forge/osx-64/libexpat-2.6.2-h73e2aa4_0.conda
  sha256: a188a77b275d61159a32ab547f7d17892226e7dac4518d2c6ac3ac8fc8dfde92
  md5: 3d1d51c8f716d97c864d12f7af329526
  constrains:
  - expat 2.6.2.*
  license: MIT
  license_family: MIT
  purls: []
  size: 69246
  timestamp: 1710362566073
- conda: https://conda.anaconda.org/conda-forge/osx-64/libffi-3.4.2-h0d85af4_5.tar.bz2
  sha256: 7a2d27a936ceee6942ea4d397f9c7d136f12549d86f7617e8b6bad51e01a941f
  md5: ccb34fb14960ad8b125962d3d79b31a9
  license: MIT
  license_family: MIT
  purls: []
  size: 51348
  timestamp: 1636488394370
- conda: https://conda.anaconda.org/conda-forge/osx-64/libsqlite-3.46.1-h4b8f8c9_0.conda
  sha256: 1d075cb823f0cad7e196871b7c57961d669cbbb6cd0e798bf50cbf520dda65fb
  md5: 84de0078b58f899fc164303b0603ff0e
  depends:
  - __osx >=10.13
  - libzlib >=1.3.1,<2.0a0
  license: Unlicense
  purls: []
  size: 908317
  timestamp: 1725353652135
- conda: https://conda.anaconda.org/conda-forge/osx-64/libzlib-1.3.1-h87427d6_1.conda
  sha256: 80a62db652b1da0ccc100812a1d86e94f75028968991bfb17f9536f3aa72d91d
  md5: b7575b5aa92108dcc9aaab0f05f2dbce
  depends:
  - __osx >=10.13
  constrains:
  - zlib 1.3.1 *_1
  license: Zlib
  license_family: Other
  purls: []
  size: 57372
  timestamp: 1716874211519
- pypi: https://files.pythonhosted.org/packages/42/d7/1ec15b46af6af88f19b8e5ffea08fa375d433c998b8a7639e76935c14f1f/markdown_it_py-3.0.0-py3-none-any.whl
  name: markdown-it-py
  version: 3.0.0
  sha256: 355216845c60bd96232cd8d8c40e8f9765cc86f46880e43a8fd22dc1a1a8cab1
  requires_dist:
  - mdurl~=0.1
  - psutil ; extra == 'benchmarking'
  - pytest ; extra == 'benchmarking'
  - pytest-benchmark ; extra == 'benchmarking'
  - pre-commit~=3.0 ; extra == 'code-style'
  - commonmark~=0.9 ; extra == 'compare'
  - markdown~=3.4 ; extra == 'compare'
  - mistletoe~=1.0 ; extra == 'compare'
  - mistune~=2.0 ; extra == 'compare'
  - panflute~=2.3 ; extra == 'compare'
  - linkify-it-py>=1,<3 ; extra == 'linkify'
  - mdit-py-plugins ; extra == 'plugins'
  - gprof2dot ; extra == 'profiling'
  - mdit-py-plugins ; extra == 'rtd'
  - myst-parser ; extra == 'rtd'
  - pyyaml ; extra == 'rtd'
  - sphinx ; extra == 'rtd'
  - sphinx-copybutton ; extra == 'rtd'
  - sphinx-design ; extra == 'rtd'
  - sphinx-book-theme ; extra == 'rtd'
  - jupyter-sphinx ; extra == 'rtd'
  - coverage ; extra == 'testing'
  - pytest ; extra == 'testing'
  - pytest-cov ; extra == 'testing'
  - pytest-regressions ; extra == 'testing'
  requires_python: '>=3.8'
- pypi: https://files.pythonhosted.org/packages/b3/38/89ba8ad64ae25be8de66a6d463314cf1eb366222074cfda9ee839c56a4b4/mdurl-0.1.2-py3-none-any.whl
  name: mdurl
  version: 0.1.2
  sha256: 84008a41e51615a49fc9966191ff91509e3c40b939176e643fd50a5c2196b8f8
  requires_python: '>=3.7'
- conda: https://conda.anaconda.org/conda-forge/osx-64/ncurses-6.5-hf036a51_1.conda
  sha256: b0b3180039ef19502525a2abd5833c00f9624af830fd391f851934d57bffb9af
  md5: e102bbf8a6ceeaf429deab8032fc8977
  depends:
  - __osx >=10.13
  license: X11 AND BSD-3-Clause
  purls: []
  size: 822066
  timestamp: 1724658603042
- conda: https://conda.anaconda.org/conda-forge/osx-64/openssl-3.3.2-hd23fc13_0.conda
  sha256: 2b75d4b56e45992adf172b158143742daeb316c35274b36f385ccb6644e93268
  md5: 2ff47134c8e292868a4609519b1ea3b6
  depends:
  - __osx >=10.13
  - ca-certificates
  license: Apache-2.0
  license_family: Apache
  purls: []
  size: 2544654
  timestamp: 1725410973572
- conda: https://conda.anaconda.org/conda-forge/noarch/packaging-24.1-pyhd8ed1ab_0.conda
  sha256: 36aca948219e2c9fdd6d80728bcc657519e02f06c2703d8db3446aec67f51d81
  md5: cbe1bb1f21567018ce595d9c2be0f0db
  depends:
  - python >=3.8
  license: Apache-2.0
  license_family: APACHE
  purls:
  - pkg:pypi/packaging?source=hash-mapping
  size: 50290
  timestamp: 1718189540074
- conda: https://conda.anaconda.org/conda-forge/noarch/pluggy-1.5.0-pyhd8ed1ab_0.conda
  sha256: 33eaa3359948a260ebccf9cdc2fd862cea5a6029783289e13602d8e634cd9a26
  md5: d3483c8fc2dc2cc3f5cf43e26d60cabf
  depends:
  - python >=3.8
  license: MIT
  license_family: MIT
  purls:
  - pkg:pypi/pluggy?source=hash-mapping
  size: 23815
  timestamp: 1713667175451
- pypi: https://files.pythonhosted.org/packages/f7/3f/01c8b82017c199075f8f788d0d906b9ffbbc5a47dc9918a945e13d5a2bda/pygments-2.18.0-py3-none-any.whl
  name: pygments
  version: 2.18.0
  sha256: b8e6aca0523f3ab76fee51799c488e38782ac06eafcf95e7ba832985c8e7b13a
  requires_dist:
  - colorama>=0.4.6 ; extra == 'windows-terminal'
  requires_python: '>=3.8'
- conda: https://conda.anaconda.org/conda-forge/noarch/pytest-8.3.2-pyhd8ed1ab_0.conda
  sha256: 72c84a3cd9fe82835a88e975fd2a0dbf2071d1c423ea4f79e7930578c1014873
  md5: e010a224b90f1f623a917c35addbb924
  depends:
  - colorama
  - exceptiongroup >=1.0.0rc8
  - iniconfig
  - packaging
  - pluggy <2,>=1.5
  - python >=3.8
  - tomli >=1
  constrains:
  - pytest-faulthandler >=2
  license: MIT
  license_family: MIT
  purls:
  - pkg:pypi/pytest?source=hash-mapping
  size: 257671
  timestamp: 1721923749407
- conda: https://conda.anaconda.org/conda-forge/osx-64/python-3.12.5-h37a9e06_0_cpython.conda
  sha256: c0f39e625b2fd65f70a9cc086fe4b25cc72228453dbbcd92cd5d140d080e38c5
  md5: 517cb4e16466f8d96ba2a72897d14c48
  depends:
  - __osx >=10.13
  - bzip2 >=1.0.8,<2.0a0
  - libexpat >=2.6.2,<3.0a0
  - libffi >=3.4,<4.0a0
  - libsqlite >=3.46.0,<4.0a0
  - libzlib >=1.3.1,<2.0a0
  - ncurses >=6.5,<7.0a0
  - openssl >=3.3.1,<4.0a0
  - readline >=8.2,<9.0a0
  - tk >=8.6.13,<8.7.0a0
  - tzdata
  - xz >=5.2.6,<6.0a0
  constrains:
  - python_abi 3.12.* *_cp312
  license: Python-2.0
  purls: []
  size: 12173272
  timestamp: 1723142761765
- conda: https://conda.anaconda.org/conda-forge/osx-64/python_abi-3.12-5_cp312.conda
  build_number: 5
  sha256: 4da26c7508d5bc5d8621e84dc510284402239df56aab3587a7d217de9d3c806d
  md5: c34dd4920e0addf7cfcc725809f25d8e
  constrains:
  - python 3.12.* *_cpython
  license: BSD-3-Clause
  license_family: BSD
  purls: []
  size: 6312
  timestamp: 1723823137004
- conda: https://conda.anaconda.org/conda-forge/osx-64/pyyaml-6.0.2-py312hb553811_1.conda
  sha256: 455ce40588b35df654cb089d29cc3f0d3c78365924ffdfc6ee93dba80cea5f33
  md5: 66514594817d51c78db7109a23ad322f
  depends:
  - __osx >=10.13
  - python >=3.12,<3.13.0a0
  - python_abi 3.12.* *_cp312
  - yaml >=0.2.5,<0.3.0a0
  license: MIT
  license_family: MIT
  purls:
  - pkg:pypi/pyyaml?source=hash-mapping
  size: 189347
  timestamp: 1725456465705
- conda: https://conda.anaconda.org/conda-forge/osx-64/readline-8.2-h9e318b2_1.conda
  sha256: 41e7d30a097d9b060037f0c6a2b1d4c4ae7e942c06c943d23f9d481548478568
  md5: f17f77f2acf4d344734bda76829ce14e
  depends:
  - ncurses >=6.3,<7.0a0
  license: GPL-3.0-only
  license_family: GPL
  purls: []
  size: 255870
  timestamp: 1679532707590
- pypi: https://files.pythonhosted.org/packages/c7/d9/c2a126eeae791e90ea099d05cb0515feea3688474b978343f3cdcfe04523/rich-13.8.0-py3-none-any.whl
  name: rich
  version: 13.8.0
  sha256: 2e85306a063b9492dffc86278197a60cbece75bcb766022f3436f567cae11bdc
  requires_dist:
  - ipywidgets>=7.5.1,<9 ; extra == 'jupyter'
  - markdown-it-py>=2.2.0
  - pygments>=2.13.0,<3.0.0
  - typing-extensions>=4.0.0,<5.0 ; python_full_version < '3.9'
  requires_python: '>=3.7.0'
- conda: https://conda.anaconda.org/conda-forge/osx-64/tk-8.6.13-h1abcd95_1.conda
  sha256: 30412b2e9de4ff82d8c2a7e5d06a15f4f4fef1809a72138b6ccb53a33b26faf5
  md5: bf830ba5afc507c6232d4ef0fb1a882d
  depends:
  - libzlib >=1.2.13,<2.0.0a0
  license: TCL
  license_family: BSD
  purls: []
  size: 3270220
  timestamp: 1699202389792
- conda: https://conda.anaconda.org/conda-forge/noarch/tomli-2.0.1-pyhd8ed1ab_0.tar.bz2
  sha256: 4cd48aba7cd026d17e86886af48d0d2ebc67ed36f87f6534f4b67138f5a5a58f
  md5: 5844808ffab9ebdb694585b50ba02a96
  depends:
  - python >=3.7
  license: MIT
  license_family: MIT
  purls:
  - pkg:pypi/tomli?source=hash-mapping
  size: 15940
  timestamp: 1644342331069
- conda: https://conda.anaconda.org/conda-forge/noarch/tzdata-2024a-h8827d51_1.conda
  sha256: 7d21c95f61319dba9209ca17d1935e6128af4235a67ee4e57a00908a1450081e
  md5: 8bfdead4e0fff0383ae4c9c50d0531bd
  license: LicenseRef-Public-Domain
  purls: []
  size: 124164
  timestamp: 1724736371498
- conda: https://conda.anaconda.org/conda-forge/osx-64/xz-5.2.6-h775f41a_0.tar.bz2
  sha256: eb09823f34cc2dd663c0ec4ab13f246f45dcd52e5b8c47b9864361de5204a1c8
  md5: a72f9d4ea13d55d745ff1ed594747f10
  license: LGPL-2.1 and GPL-2.0
  purls: []
  size: 238119
  timestamp: 1660346964847
- conda: https://conda.anaconda.org/conda-forge/osx-64/yaml-0.2.5-h0d85af4_2.tar.bz2
  sha256: 5301417e2c8dea45b401ffee8df3957d2447d4ce80c83c5ff151fc6bfe1c4148
  md5: d7e08fcf8259d742156188e8762b4d20
  license: MIT
  license_family: MIT
  purls: []
  size: 84237
  timestamp: 1641347062780
